### PR TITLE
Improve support for program files omitting the `program` header

### LINKED ISSUE #227: Error in symbol table construction for files without the `program` header
### Prerequisites

- [X] This bug is in SonarDelphi, not SonarQube or my Delphi code.
- [X] This bug has not already been reported.

### SonarDelphi version

1.4.0

### SonarQube version

_No response_

### Issue description

The `program <name>` header isn't technically necessary in a Delphi program file.
The compiler will happily accept
```delphi
begin
end.
```

but scanning with sonar-delphi gives the following error
```
java.lang.ClassCastException: class au.com.integradev.delphi.antlr.ast.node.CommonDelphiNodeImpl cannot be cast to class org.sonar.plugins.communitydelphi.api.ast.FileHeaderNode (au.com.integradev.delphi.antlr.ast.node.CommonDelphiNodeImpl and org.sonar.plugins.communitydelphi.api.ast.FileHeaderNode are in unnamed module of loader org.sonar.classloader.ClassRealm @3549bca9)
	at au.com.integradev.delphi.antlr.ast.DelphiAstImpl.getFileHeader(DelphiAstImpl.java:96)
	at au.com.integradev.delphi.antlr.ast.visitors.SymbolTableVisitor$1.visit(SymbolTableVisitor.java:246)
	at au.com.integradev.delphi.symbol.SymbolTableBuilder.runSymbolTableVisitor(SymbolTableBuilder.java:315)
	at au.com.integradev.delphi.symbol.SymbolTableBuilder.process(SymbolTableBuilder.java:287)
	at au.com.integradev.delphi.symbol.SymbolTableBuilder.indexUnit(SymbolTableBuilder.java:396)
	at au.com.integradev.delphi.symbol.SymbolTableBuilder.build(SymbolTableBuilder.java:477)
	at au.com.integradev.delphi.DelphiSensor.executeOnFiles(DelphiSensor.java:119)
	at au.com.integradev.delphi.DelphiSensor.execute(DelphiSensor.java:83)
	at org.sonarsource.sonarlint.core.analysis.container.analysis.sensor.SensorsExecutor.executeSensor(SensorsExecutor.java:75)
	at org.sonarsource.sonarlint.core.analysis.container.analysis.sensor.SensorsExecutor.execute(SensorsExecutor.java:66)
	at org.sonarsource.sonarlint.core.analysis.container.analysis.AnalysisContainer.doAfterStart(AnalysisContainer.java:122)
	at org.sonarsource.sonarlint.core.plugin.commons.container.SpringComponentContainer.startComponents(SpringComponentContainer.java:182)
	at org.sonarsource.sonarlint.core.plugin.commons.container.SpringComponentContainer.execute(SpringComponentContainer.java:161)
	at org.sonarsource.sonarlint.core.analysis.container.module.ModuleContainer.analyze(ModuleContainer.java:71)
	at au.com.integradev.delphilint.analysis.AnalysisOrchestrator.runAnalysis(AnalysisOrchestrator.java:147)
	at au.com.integradev.delphilint.server.AnalysisServer.analyze(AnalysisServer.java:115)
	at au.com.integradev.delphilint.server.TlvConnection.processRequest(TlvConnection.java:171)
	at au.com.integradev.delphilint.server.TlvConnection.readStream(TlvConnection.java:158)
	at au.com.integradev.delphilint.server.TlvConnection.run(TlvConnection.java:70)
	at au.com.integradev.delphilint.App.main(App.java:70)
```

It seems that the symbol table construction code is assuming what kind of node the AST starts with.

### Steps to reproduce

Analyse the following code with sonar-delphi
```delphi
begin
end.
```

### Minimal Delphi code exhibiting the issue

_No response_

### PR DESCRIPTION
**Pull request by Cirras**
This was partially implemented in #159, but stopped at the parsing level.

Throughout the analyzer, there were unsafe assumptions about the presence of a file header. The major area that needed extra consideration was symbol table construction.

It turns out that the default name of a Delphi program is `PROGRAM`, so this name ends up being used if the header is omitted. See `DefaultProgramName.dpr` for an example of a valid Delphi program that relies on this behavior.

Fixes #227

### CODE REVIEW

**Comment by zaneduffield** [comment]
```delphi
const
  HelloWorld = 'Hello, World!';

begin
  WriteLn(&PROGRAM.HelloWorld);
end.
```
cursed program

File: delphi-frontend/src/main/java/au/com/integradev/delphi/symbol/declaration/QualifiedNameDeclarationImpl.java [hunk 1]
```diff
@@ -30,13 +32,18 @@ public abstract class QualifiedNameDeclarationImpl extends NameDeclarationImpl
   private final QualifiedName qualifiedName;
 
   protected QualifiedNameDeclarationImpl(QualifiedNameDeclarationNode node) {
-    super(node);
-    qualifiedName = node.getQualifiedName();
+    super(node, node.getScope());
+    this.qualifiedName = node.getQualifiedName();
```
Comment (fourls) [inline delphi-frontend/src/main/java/au/com/integradev/delphi/symbol/declaration/QualifiedNameDeclarationImpl.java:36]: Is this change necessary? It looks like it's functionally equivalent to how it was before.

**Review by fourls — CHANGES_REQUESTED** [review]
Looks good - for the most part, if the tests and our QA are happy with it, I'm happy with it. Just one nitpick.

Also, were you intending to also add support for the name of the program begin/end block being the same name as the program file? I think it would be a suitable change to make either with this PR or soon after.

File: delphi-frontend/src/main/java/au/com/integradev/delphi/symbol/declaration/QualifiedNameDeclarationImpl.java (hunk 1 above)
Comment (Cirras) [reply delphi-frontend/src/main/java/au/com/integradev/delphi/symbol/declaration/QualifiedNameDeclarationImpl.java:36]: Oops, artifact of a change I almost made and later decided against.
I'll revert that.

**Comment by Cirras** [comment]
> Also, were you intending to also add support for the name of the program begin/end block being the same name as the program file? I think it would be a suitable change to make either with this PR or soon after.

Can you elaborate on this point a bit?
The debugger does show `<program_name.program_name>` for the entry point, but the entry point is not a function and cannot be invoked. I don't really see how we'd model that information or surface it in a useful way.

**Review by fourls — APPROVED** [review]
Looks good.

**Comment by fourls** [comment]
> > Also, were you intending to also add support for the name of the program begin/end block being the same name as the program file? I think it would be a suitable change to make either with this PR or soon after.
> 
> Can you elaborate on this point a bit? The debugger does show `<program_name.program_name>` for the entry point, but the entry point is not a function and cannot be invoked. I don't really see how we'd model that information or surface it in a useful way.

My mistake, I thought the entry point could be invoked. No worries then.
